python/taichi/lang/transformer.py

```diff
@@ -278,64 +278,108 @@ def visit_For(self, node):
         if node.orelse:
             raise TaichiSyntaxError(
                 "'else' clause for 'for' not supported in Taichi kernels")
-        decorated = isinstance(node.iter, ast.Call) and isinstance(
-            node.iter.func, ast.Attribute) and isinstance(node.iter.func.value, ast.Name) \
-                    and node.iter.func.value.id == 'ti'
-        is_ndrange_for = False
-        is_static_for = False
-        is_grouped = False
+        decorated = isinstance(node.iter, ast.Call) \
```

Comment:
These lines are way too long - maybe it's worth extracting some common patterns into a (static) member method :-)